File: skills/install-slop-guard/assets/python/rules/smells/no_temporary_field.py
"""Reject attributes that only exist for part of an object's life."""

from __future__ import annotations

import ast

from ...rule import Rule


class NoTemporaryField(Rule):
    name = "no-temporary-field"
    group = "smells"
    reference = "https://refactoring.guru/smells/temporary-field"
    description = "Disallow `self` attributes first assigned outside `__init__`."

    def visit_ClassDef(self, node: ast.ClassDef) -> None:
        declared = {
            target.id
            for statement in node.body
            if isinstance(statement, ast.AnnAssign) and isinstance(statement.target, ast.Name)
            for target in [statement.target]
        }
        for statement in node.body:
            if isinstance(statement, ast.Assign):
                declared.update(
                    target.id for target in statement.targets if isinstance(target, ast.Name)
                )
        initialised: set[str] = set()
        for statement in node.body:
            if isinstance(statement, (ast.FunctionDef, ast.AsyncFunctionDef)) and statement.name == "__init__":
                initialised = self._assigned(statement)
        for statement in node.body:
            if not isinstance(statement, (ast.FunctionDef, ast.AsyncFunctionDef)):
                continue
            if statement.name == "__init__":
                continue
            for attribute in sorted(self._assigned(statement) - initialised - declared):
                self.report(
                    statement,
                    f"`self.{attribute}` is created in `{statement.name}` rather than in "
                    f"`__init__`, so between construction and that call the attribute does not "
                    f"exist and every reader has to work out when it does. Set it in `__init__`, "
                    f"or pass it between the methods that use it.",
                )
        self.generic_visit(node)

    @staticmethod
    def _assigned(node: ast.AST) -> set[str]:
        return {
            child.attr
            for child in ast.walk(node)
            if isinstance(child, ast.Attribute)
            and isinstance(child.ctx, ast.Store)
            and isinstance(child.value, ast.Name)
            and child.value.id == "self"
        }

File: languages/typescript/src/smells/rules/no-temporary-field.ts
import { defineRule } from "@oxlint/plugins";
import type { ESTree } from "@oxlint/plugins";

import { walk } from "../shared/traversal.ts";

function assignedFields(node: ESTree.Node, visitorKeys: Readonly<Record<string, readonly string[]>>): Set<string> {
	const found = new Set<string>();
	walk(node, visitorKeys, (child) => {
		if (
			child.type === "AssignmentExpression" &&
			child.left.type === "MemberExpression" &&
			child.left.object.type === "ThisExpression" &&
			child.left.property.type === "Identifier"
		) {
			found.add(child.left.property.name);
		}
	});
	return found;
}

/** Ban fields that only exist for part of an object's life. */
export const noTemporaryFieldRule = defineRule({
	meta: {
		type: "problem",
		docs: {
			description:
				"Disallow `this` fields first assigned outside the constructor and never declared on the class.",
		},
		messages: {
			temporaryField:
				"`this.{{field}}` is created in `{{method}}` rather than in the constructor, so between construction and that call the field does not exist and every reader has to work out when it does. Declare it on the class, or pass it between the methods that use it.",
		},
	},
	createOnce(context) {
		return {
			ClassBody(node) {
				const declared = new Set<string>();
				for (const member of node.body) {
					if (
						(member.type === "PropertyDefinition" || member.type === "AccessorProperty") &&
						member.key.type === "Identifier"
					) {
						declared.add(member.key.name);
					}
				}
				const constructor = node.body.find(
					(member) => member.type === "MethodDefinition" && member.kind === "constructor",
				);
				const initialised =
					constructor === undefined
						? new Set<string>()
						: assignedFields(constructor, context.sourceCode.visitorKeys);

				for (const member of node.body) {
					if (member.type !== "MethodDefinition" || member.kind === "constructor") continue;
					const method =
						member.key.type === "Identifier" ? member.key.name : context.sourceCode.getText(member.key);
					for (const field of assignedFields(member, context.sourceCode.visitorKeys)) {
						if (declared.has(field) || initialised.has(field)) continue;
						context.report({
							node: member,
							messageId: "temporaryField",
							data: { field, method },
						});
					}
				}
			},
		};
	},
});
